src/main/java/com/google/sps/data/DatastoreAvailabilityDao.java

```diff
@@ -0,0 +1,176 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/** Accesses Datastore to support managing Availability entities. */
+public class DatastoreAvailabilityDao implements AvailabilityDao {
+  // @param datastore the DatastoreService we're using to interact with Datastore.
+  private DatastoreService datastore;
+
+  /** Initializes the fields for DatastoreAvailabilityDao. */
+  public DatastoreAvailabilityDao() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /**
+   * Retrieve the availability from Datastore from its id and wrap it in an Optional. If the id
+   * isn't in Datastore, the Optional is empty.
+   */
+  @Override
+  public Optional<Availability> get(long id) {
+    Key key = KeyFactory.createKey("Availability", id);
+    Entity availabilityEntity;
+    try {
+      availabilityEntity = datastore.get(key);
+    } catch (com.google.appengine.api.datastore.EntityNotFoundException e) {
+      return Optional.empty();
+    }
+    return Optional.of(entityToAvailability(availabilityEntity));
+  }
+
+  // Adds an Availability object into Datastore.
+  @Override
+  public void create(Availability avail) {
+    datastore.put(availabilityToNewEntity(avail));
+  }
+
+  // Updates the specified id with the new availability.
+  @Override
+  public void update(Availability avail) {
+    datastore.put(availabilityToUpdatedEntity(avail));
+  }
+
+  static Entity availabilityToNewEntity(Availability avail) {
+    Entity availabilityEntity = new Entity("Availability");
+    setEntityProperties(availabilityEntity, avail);
+    return availabilityEntity;
+  }
+
+  static Entity availabilityToUpdatedEntity(Availability avail) {
+    Entity availabilityEntity = new Entity("Availability", avail.id());
+    setEntityProperties(availabilityEntity, avail);
+    return availabilityEntity;
+  }
+
+  static void setEntityProperties(Entity availabilityEntity, Availability avail) {
+    availabilityEntity.setProperty("email", avail.email());
+    availabilityEntity.setProperty("startTime", avail.when().start().toEpochMilli());
+    availabilityEntity.setProperty("endTime", avail.when().end().toEpochMilli());
+    availabilityEntity.setProperty("scheduled", avail.scheduled());
+    return;
+  }
+
+  static Availability entityToAvailability(Entity availabilityEntity) {
+    return Availability.create(
+        (String) availabilityEntity.getProperty("email"),
+        TimeRange.fromStartEnd(
+            Instant.ofEpochMilli((long) availabilityEntity.getProperty("startTime")),
+            Instant.ofEpochMilli((long) availabilityEntity.getProperty("endTime"))),
+        availabilityEntity.getKey().getId(),
+        (boolean) availabilityEntity.getProperty("scheduled"));
+  }
+
+  // Deletes all Availability entities for a user ranging from minTime to maxTime.
+  // minTime and maxTime are in milliseconds.
+  public void deleteInRangeForUser(String email, long minTime, long maxTime) {
```

Comment:
Shouldn't this method include an @Override annotation ?

Reply:
Done